File: frontend/src/services/flow.ts
/**
 * Flow Service - API client for Flow Builder operations
 */
import { api } from './api'

// ============================================================================
// ENUMS
// ============================================================================

export enum NodeType {
  SOURCE = 'source',
  CHECK = 'check',
  JOIN = 'join',
  FILTER = 'filter',
  AGGREGATE = 'aggregate',
  TRANSFORM = 'transform',
}

export enum FlowStatus {
  DRAFT = 'draft',
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  ARCHIVED = 'archived',
}

export enum ExecutionStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  WARNING = 'warning',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export enum ExecutionTrigger {
  MANUAL = 'manual',
  SCHEDULED = 'scheduled',
  API = 'api',
}

// ============================================================================
// INTERFACES
// ============================================================================

export interface NodePosition {
  x: number
  y: number
}

export interface FlowNode {
  id: string
  type: NodeType
  label: string
  checkType?: string
  position: NodePosition
  config: Record<string, any>
}

export interface FlowConnection {
  id: string
  from: string
  to: string
  source_output?: string
  target_input?: string
}

export interface FlowDefinition {
  nodes: FlowNode[]
  connections: FlowConnection[]
  metadata?: Record<string, any>
}

export interface FlowSchedule {
  enabled: boolean
  cron_expression?: string
  timezone?: string
}

export interface Flow {
  id: string
  workspace_id: string
  name: string
  description?: string
  flow_definition: FlowDefinition
  status: FlowStatus
  created_by: string
  owner_user_id?: string | null
  created_at: string
  updated_at: string
  last_executed_at?: string
  schedule?: FlowSchedule
  tags?: string[]
  version: number
  execution_count: number
}

export interface FlowExecution {
  id: string
  flow_id: string
  flow_version: number
  status: ExecutionStatus
  trigger: ExecutionTrigger
  triggered_by?: string
  started_at?: string
  completed_at?: string
  duration_seconds?: number
  execution_time_seconds?: number  // alias kept for backward compat
  nodes_executed: number
  nodes_passed: number
  nodes_failed: number
  nodes_skipped: number
  result_summary?: Record<string, any>
  error_message?: string
  execution_config?: Record<string, any>
  // Enhanced fields from detail endpoint
  flow_name?: string
  executed_by_name?: string
}

export interface FlowNodeResult {
  id: string
  execution_id: string
  node_id: string
  status: ExecutionStatus
  execution_order: number
  started_at?: string
  completed_at?: string
  execution_time_seconds?: number
  result_data?: Record<string, any>
  error_message?: string
}

export interface FlowTemplate {
  id: string
  name: string
  description?: string
  category?: string
  template_definition: FlowDefinition
  preview_image_url?: string
  created_by: string
  created_at: string
  updated_at: string
  use_count: number
  is_public: boolean
  tags?: string[]
}

export interface ValidationError {
  node_id?: string
  connection_id?: string
  error_type: string
  message: string
}

export interface FlowValidationResponse {
  is_valid: boolean
  errors: ValidationError[]
  warnings: ValidationError[]
  execution_order?: string[]
}

// ============================================================================
// REQUEST/RESPONSE TYPES
// ============================================================================

export interface CreateFlowRequest {
  name: string
  description?: string
  flow_definition: FlowDefinition
  status?: FlowStatus
  schedule?: FlowSchedule
  tags?: string[]
}

export interface UpdateFlowRequest {
  name?: string
  description?: string
  flow_definition?: FlowDefinition
  status?: FlowStatus
  schedule?: FlowSchedule
  tags?: string[]
}

export interface ListFlowsParams {
  status?: FlowStatus
  tags?: string[]
  search?: string
  skip?: number
  limit?: number
}

export interface ExecuteFlowRequest {
  execution_config?: Record<string, any>
}

export interface ExportFlowRequest {
  format: 'json' | 'yaml'
  include_metadata?: boolean
}

export interface ImportFlowRequest {
  data: string
  format: 'json' | 'yaml'
}

export interface DuplicateFlowRequest {
  new_name: string
  copy_schedule?: boolean
}

// ============================================================================
// FLOW SERVICE CLASS
// ============================================================================

class FlowService {
  private baseUrl = '/workspaces'

  /**
   * Create a new flow
   */
  async createFlow(workspaceId: string, data: CreateFlowRequest): Promise<Flow> {
    const response = await api.post(`${this.baseUrl}/${workspaceId}/flows`, data)
    return response.data
  }

  /**
   * Get a flow by ID
   */
  async getFlow(workspaceId: string, flowId: string): Promise<Flow> {
    const response = await api.get(`${this.baseUrl}/${workspaceId}/flows/${flowId}`)
    return response.data
  }

  /**
   * List flows with optional filters
   */
  /**
   * List flows with optional filters
   */
  async listFlows(workspaceId: string, params?: ListFlowsParams): Promise<Flow[]> {
    const response = await api.get(`${this.baseUrl}/${workspaceId}/flows`, { params })
    // Backend returns { flows: [], total, page, page_size }
    return response.data.flows || []
  }

  /**
   * Update a flow
   */
  async updateFlow(
    workspaceId: string,
    flowId: string,
    data: UpdateFlowRequest
  ): Promise<Flow> {
    const response = await api.patch(`${this.baseUrl}/${workspaceId}/flows/${flowId}`, data)
    return response.data
  }

  /**
   * Delete a flow
   */
  async deleteFlow(workspaceId: string, flowId: string, hardDelete: boolean = true): Promise<void> {
    await api.delete(`${this.baseUrl}/${workspaceId}/flows/${flowId}?hard_delete=${hardDelete}`)
  }

  /**
   * Assign or clear the owner of a flow
   */
  async assignOwner(
    workspaceId: string,
    flowId: string,
    ownerUserId: string | null,
  ): Promise<{ flow_id: string; owner_user_id: string | null; previous_owner_user_id: string | null }> {
    const { data } = await api.put(
      `${this.baseUrl}/${workspaceId}/flows/${flowId}/owner`,
      { owner_user_id: ownerUserId },
    )
    return data
  }

  /**
   * Validate a flow definition
   */
  async validateFlow(
    workspaceId: string,
    flowDefinition: FlowDefinition
  ): Promise<FlowValidationResponse> {
    const response = await api.post(
      `${this.baseUrl}/${workspaceId}/flows/validate`,
      { flow_definition: flowDefinition }
    )
    return response.data
  }

  /**
   * Execute a flow
   */
  async executeFlow(
    workspaceId: string,
    flowId: string,
    config?: ExecuteFlowRequest
  ): Promise<FlowExecution> {
    const response = await api.post(
      `${this.baseUrl}/${workspaceId}/flows/${flowId}/execute`,
      config || {}
    )
    return response.data
  }

  /**
   * Get flow executions
   */
  async getFlowExecutions(
    workspaceId: string,
    flowId: string,
    limit?: number
  ): Promise<FlowExecution[]> {
    const response = await api.get(
      `${this.baseUrl}/${workspaceId}/flows/${flowId}/executions`,
      { params: { limit } }
    )
    // Backend returns { executions: [], total, page, page_size }
    return response.data.executions || []
  }

  /**
   * Get execution details
   */
  async getExecution(
    workspaceId: string,
    executionId: string
  ): Promise<FlowExecution> {
    const response = await api.get(
      `${this.baseUrl}/${workspaceId}/flow-executions/${executionId}`
    )
    return response.data
  }

  /**
   * List all executions for a workspace (across all flows)
   */
  async listAllExecutions(
    workspaceId: string,
    params?: { status?: string; page?: number; page_size?: number }
  ): Promise<FlowExecution[]> {
    const response = await api.get(
      `${this.baseUrl}/${workspaceId}/flow-executions`,
      { params }
    )
    return response.data.executions || []
  }

  /**
   * Get node results for an execution
   */
  async getNodeResults(
    workspaceId: string,
    executionId: string
  ): Promise<FlowNodeResult[]> {
    const response = await api.get(
      `${this.baseUrl}/${workspaceId}/flow-executions/${executionId}/nodes`
    )
    return response.data
  }

  /**
   * Cancel a running execution
   */
  async cancelExecution(workspaceId: string, executionId: string): Promise<void> {
    await api.delete(`${this.baseUrl}/${workspaceId}/flow-executions/${executionId}`)
  }

  /**
   * Export a flow
   */
  async exportFlow(
    workspaceId: string,
    flowId: string,
    format: 'json' | 'yaml' = 'json'
  ): Promise<string> {
    const response = await api.post(
      `${this.baseUrl}/${workspaceId}/flows/${flowId}/export`,
      { format },
      { responseType: 'text' }
    )
    return response.data
  }

  /**
   * Import a flow
   */
  async importFlow(
    workspaceId: string,
    data: string,
    format: 'json' | 'yaml' = 'json'
  ): Promise<Flow> {
    const response = await api.post(`${this.baseUrl}/${workspaceId}/flows/import`, {
      data,
      format,
    })
    return response.data
  }

  /**
   * Duplicate a flow
   */
  async duplicateFlow(
    workspaceId: string,
    flowId: string,
    newName: string,
    copySchedule: boolean = false
  ): Promise<Flow> {
    const response = await api.post(
      `${this.baseUrl}/${workspaceId}/flows/${flowId}/duplicate`,
      { new_name: newName, copy_schedule: copySchedule }
    )
    return response.data
  }

  /**
   * List flow templates
   */
  async listTemplates(workspaceId: string): Promise<FlowTemplate[]> {
    const response = await api.get(`${this.baseUrl}/${workspaceId}/templates`)
    return response.data
  }

  /**
   * Create flow from template
   */
  async createFromTemplate(
    workspaceId: string,
    templateId: string,
    flowName: string
  ): Promise<Flow> {
    const response = await api.post(
      `${this.baseUrl}/${workspaceId}/templates/${templateId}/create-flow`,
      { name: flowName }
    )
    return response.data
  }

  /**
   * AI-powered flow building from natural language prompts
   */
  async aiBuildFlow(
    workspaceId: string,
    prompt: string,
    currentFlow: { nodes: FlowNode[], connections: FlowConnection[] },
    availableDataSources?: any[]
  ): Promise<{
    success: boolean
    needs_clarification: boolean
    clarification_questions: string[]
    suggested_data_sources?: any[]
    pending_tasks?: any[]
    flow_updates: { nodes: FlowNode[], connections: FlowConnection[] } | null
    message: string
  }> {
    const response = await api.post(
      `${this.baseUrl}/${workspaceId}/flows/ai-build`,
      { 
        prompt, 
        current_flow: currentFlow,
        available_data_sources: availableDataSources || []
      }
    )
    return response.data
  }

  // ── F115 — Schedule Management ───────────────────────────────────

  async getSchedule(workspaceId: string, flowId: string) {
    const response = await api.get(`${this.baseUrl}/${workspaceId}/flows/${flowId}/schedule`)
    return response.data as { flow_id: string; schedule: FlowSchedule | null; next_run_at: string | null }
  }

  async setSchedule(workspaceId: string, flowId: string, schedule: { enabled: boolean; cron: string; timezone?: string }) {
    const response = await api.put(`${this.baseUrl}/${workspaceId}/flows/${flowId}/schedule`, schedule)
    return response.data
  }

  async removeSchedule(workspaceId: string, flowId: string) {
    const response = await api.delete(`${this.baseUrl}/${workspaceId}/flows/${flowId}/schedule`)
    return response.data
  }

  async listScheduledFlows(workspaceId: string) {
    const response = await api.get(`${this.baseUrl}/${workspaceId}/flows/schedules`)
    return response.data as { schedules: Array<{ flow_id: string; flow_name: string; schedule: FlowSchedule; next_run_at: string | null; status: string }>; total: number }
  }

  async validateCron(workspaceId: string, flowId: string, schedule: { enabled: boolean; cron: string; timezone?: string }) {
    const response = await api.post(`${this.baseUrl}/${workspaceId}/flows/${flowId}/schedule/validate`, schedule)
    return response.data as { valid: boolean; error?: string; next_runs: string[] }
  }
}

export default new FlowService()

File: backend/app/schemas/flow.py
"""
Flow Schemas - Pydantic models for request/response validation

This module defines the Pydantic schemas for flow-related API operations.
"""

from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

# ==============================================
# Enums
# ==============================================


class FlowStatus(str, Enum):
    """Flow status enumeration"""

    DRAFT = "draft"
    ACTIVE = "active"
    INACTIVE = "inactive"
    ARCHIVED = "archived"


class ExecutionType(str, Enum):
    """Execution type enumeration"""

    MANUAL = "manual"
    SCHEDULED = "scheduled"
    TRIGGERED = "triggered"
    TEST = "test"


# Alias for frontend compatibility
ExecutionTrigger = ExecutionType


class ExecutionStatus(str, Enum):
    """Execution status enumeration"""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class NodeStatus(str, Enum):
    """Node execution status enumeration"""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    WARNING = "warning"
    FAILED = "failed"
    SKIPPED = "skipped"


class NodeType(str, Enum):
    """Node type enumeration"""

    SOURCE = "source"
    CHECK = "check"
    JOIN = "join"
    FILTER = "filter"
    AGGREGATE = "aggregate"
    TRANSFORM = "transform"


# ==============================================
# Flow Definition Sub-schemas
# ==============================================


class NodePosition(BaseModel):
    """Node position on canvas"""

    x: float
    y: float


class FlowNode(BaseModel):
    """Flow node definition"""

    id: str = Field(..., description="Unique node identifier")
    type: NodeType = Field(..., description="Node type")
    label: str | None = Field(None, description="Node display label")
    config: dict[str, Any] = Field(default_factory=dict, description="Node configuration")
    position: NodePosition | None = Field(None, description="Node position on canvas")

    # For check nodes
    checkType: str | None = Field(None, description="Type of check (for check nodes)")

    # Visual styling
    style: dict[str, Any] | None = Field(None, description="Node visual style")


class FlowConnection(BaseModel):
    """Flow connection between nodes"""

    id: str = Field(..., description="Unique connection identifier")
    source: str = Field(..., alias="from", description="Source node ID")
    target: str = Field(..., alias="to", description="Target node ID")
    sourcePort: str | None = Field("output", description="Source port name")
    targetPort: str | None = Field("input", description="Target port name")
    label: str | None = Field(None, description="Connection label")

    class Config:
        populate_by_name = True  # Allow both 'source' and 'from'


class FlowDefinition(BaseModel):
    """Complete flow definition"""

    nodes: list[FlowNode] = Field(..., description="List of flow nodes")
    connections: list[FlowConnection] = Field(..., description="List of connections between nodes")
    metadata: dict[str, Any] | None = Field(default_factory=dict, description="Additional metadata")


class ScheduleConfig(BaseModel):
    """Flow schedule configuration"""

    enabled: bool = Field(False, description="Whether scheduling is enabled")
    cron: str = Field(..., description="Cron expression for schedule")
    timezone: str = Field("UTC", description="Timezone for schedule")


class ExecutionConfig(BaseModel):
    """Execution configuration"""

    sample_size: int | None = Field(None, description="Number of rows to sample")
    parallel: bool = Field(True, description="Execute independent nodes in parallel")
    continue_on_error: bool = Field(False, description="Continue execution if a node fails")
    timeout_seconds: int | None = Field(None, description="Execution timeout in seconds")


# ==============================================
# Request Schemas
# ==============================================


class CreateFlowRequest(BaseModel):
    """Request schema for creating a flow"""

    name: str = Field(..., min_length=1, max_length=255, description="Flow name")
    description: str | None = Field(None, description="Flow description")
    flow_definition: FlowDefinition = Field(
        ..., description="Flow definition (nodes and connections)"
    )
    status: FlowStatus = Field(FlowStatus.DRAFT, description="Initial flow status")
    schedule: ScheduleConfig | None = Field(None, description="Schedule configuration")
    tags: list[str] | None = Field(None, description="Flow tags")


class UpdateFlowRequest(BaseModel):
    """Request schema for updating a flow"""

    name: str | None = Field(None, min_length=1, max_length=255, description="Flow name")
    description: str | None = Field(None, description="Flow description")
    flow_definition: FlowDefinition | None = Field(None, description="Flow definition")
    status: FlowStatus | None = Field(None, description="Flow status")
    is_active: bool | None = Field(None, description="Whether flow is active")
    schedule: ScheduleConfig | None = Field(None, description="Schedule configuration")
    tags: list[str] | None = Field(None, description="Flow tags")


class ExecuteFlowRequest(BaseModel):
    """Request schema for executing a flow"""

    execution_config: ExecutionConfig | None = Field(None, description="Execution configuration")
    parameters: dict[str, Any] | None = Field(None, description="Runtime parameters")


class ValidateFlowRequest(BaseModel):
    """Request schema for validating a flow definition"""

    flow_definition: FlowDefinition = Field(..., description="Flow definition to validate")


class ImportFlowRequest(BaseModel):
    """Request schema for importing a flow"""

    name: str = Field(..., description="Flow name")
    description: str | None = Field(None, description="Flow description")
    flow_definition: FlowDefinition = Field(..., description="Flow definition")
    tags: list[str] | None = Field(None, description="Flow tags")


# ==============================================
# Response Schemas
# ==============================================


class FlowResponse(BaseModel):
    """Response schema for a flow"""

    id: UUID
    workspace_id: UUID
    name: str
    description: str | None
    flow_definition: dict[str, Any]  # FlowDefinition as dict
    status: str
    is_active: bool
    schedule: dict[str, Any] | None  # ScheduleConfig as dict
    tags: list[str] | None
    version: int
    created_by: UUID | None
    owner_user_id: UUID | None = None
    created_at: datetime
    updated_at: datetime

    # Optional execution statistics
    execution_count: int | None = None
    last_execution: datetime | None = None
    last_execution_status: str | None = None

    class Config:
        from_attributes = True


class FlowListResponse(BaseModel):
    """Response schema for list of flows"""

    flows: list[FlowResponse]
    total: int
    page: int
    page_size: int


class NodeResultData(BaseModel):
    """Node execution result data"""

    rows_scanned: int | None = None
    rows_passed: int | None = None
    rows_failed: int | None = None
    pass_rate: float | None = None
    violations: list[dict[str, Any]] | None = None

    # Additional fields depending on node type
    extra: dict[str, Any] | None = None


class FlowNodeResultResponse(BaseModel):
    """Response schema for a flow node result"""

    id: UUID
    execution_id: UUID
    node_id: str
    node_type: str
    status: str
    started_at: datetime | None
    completed_at: datetime | None
    duration_seconds: int | None
    result_data: dict[str, Any] | None
    error_message: str | None
    error_details: dict[str, Any] | None
    execution_order: int | None
    created_at: datetime

    class Config:
        from_attributes = True


class FlowExecutionResponse(BaseModel):
    """Response schema for a flow execution"""

    id: UUID
    flow_id: UUID
    execution_type: str = Field(..., alias="trigger", serialization_alias="trigger")
    status: str
    started_at: datetime | None
    completed_at: datetime | None
    duration_seconds: int | None
    nodes_executed: int
    nodes_passed: int
    nodes_failed: int
    nodes_skipped: int
    execution_config: dict[str, Any] | None
    result_summary: dict[str, Any] | None
    error_message: str | None
    error_details: dict[str, Any] | None
    executed_by: UUID | None
    created_at: datetime

    # Optional flow details
    flow_name: str | None = None

    # Optional node results
    node_results: list[FlowNodeResultResponse] | None = None

    # Enhanced fields for new report
    executed_by_name: str | None = None

    class Config:
        from_attributes = True
        populate_by_name = True


class FlowExecutionListResponse(BaseModel):
    """Response schema for list of flow executions"""

    executions: list[FlowExecutionResponse]
    total: int
    page: int
    page_size: int


class ValidationError(BaseModel):
    """Validation error details"""

    type: str
    message: str
    node_id: str | None = None
    connection_id: str | None = None


class FlowValidationResponse(BaseModel):
    """Response schema for flow validation"""

    is_valid: bool
    errors: list[ValidationError] = Field(default_factory=list)
    warnings: list[ValidationError] = Field(default_factory=list)

    # Validation metadata
    node_count: int
    connection_count: int
    has_source: bool
    has_checks: bool
    has_circular_dependencies: bool


class FlowExportResponse(BaseModel):
    """Response schema for flow export"""

    flow_id: UUID
    name: str
    description: str | None
    flow_definition: dict[str, Any]
    exported_at: datetime
    export_format: str  # json, yaml

    class Config:
        from_attributes = True


class FlowTemplateResponse(BaseModel):
    """Response schema for a flow template"""

    id: UUID
    name: str
    description: str | None
    category: str | None
    template_definition: dict[str, Any]
    preview_image_url: str | None
    is_public: bool
    use_count: int
    created_by: UUID | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ExecutionSummaryResponse(BaseModel):
    """Summary statistics for flow executions"""

    total_executions: int
    successful_executions: int
    failed_executions: int
    running_executions: int
    average_duration_seconds: float | None
    success_rate: float
    last_execution: FlowExecutionResponse | None
